### ISSUE TITLE
High xorg CPU use, revert to software rendering in log even with kernel-latest

### ISSUE THREAD

**Issue by weissjeffm** [QubesOS/qubes-issues#5244]
**Qubes OS version**
R4.0

**Affected component(s) or functionality**
Graphics driver

**Brief summary**
When I first installed Qubes, I had precisely the issue discussed here: https://www.qubes-os.org/doc/intel-igfx-troubleshooting/

I then installed the `kernel-latest` package and rebooted, which did improve things - the visibly slow rendering problem went away (where you can actually see painting being done row by row) and the OS is actually usable. 

However, I still see the error in the Xorg logs, and CPU use of Xorg during video playback is unusually high (far higher than it was on the same hardware when I ran fedora 30 as the base os).

I did a little digging and it seems like updating the mesa driver packages might fix this, but they're fedora-25 era and I'm not sure if it's possible to install say, the fedora-30 versions without a massive dependency chain. I tried to do this but could not figure out how to import the fedora 30 repo keys.

I realize this is relatively new hardware and dom0 is relatively old drivers, so I may just have to wait, but any advice on how to get the video drivers working (hacky or not) would be greatly appreciated.

**To Reproduce**
Boot the system, view Xorg log. Play youtube video

**Expected behavior**
CPU use should be more like 25-50%, no error in logs

**Actual behavior**
Error in Xorg log, 100%+ CPU use, high fan speed.

**Additional context**
Thinkpad T590 with Intel graphics 

lspci
```
...
00:02.0 VGA compatible controller: Intel Corporation Device 3ea0
```
kernel: `5.1.15-1.pvops.qubes.x86_64`

Xorg log snippet
```
[    54.264] (II) intel(0): SNA initialized with disabled backend
[    54.264] (==) intel(0): Backing store enabled
[    54.264] (==) intel(0): Silken mouse enabled
[    54.265] (II) intel(0): HW Cursor enabled
[    54.265] (II) intel(0): RandR 1.2 enabled, ignore the following RandR disabled message.
[    54.266] (==) intel(0): DPMS enabled
[    54.267] (==) intel(0): Display hotplug detection enabled
[    54.267] (II) intel(0): Textured video not supported on this hardware or backend
[    54.267] (WW) intel(0): loading DRI2 whilst acceleration is disabled.
[    54.267] (II) intel(0): [DRI2] Setup complete
[    54.267] (II) intel(0): [DRI2]   DRI driver: i965
[    54.267] (II) intel(0): [DRI2]   VDPAU driver: va_gl
[    54.267] (II) intel(0): direct rendering: DRI2 DRI3 enabled
[    54.267] (II) intel(0): hardware support for Present enabled
[    54.267] (--) RandR disabled
[    54.270] (II) SELinux: Disabled on system
[    54.285] (EE) AIGLX error: Calling driver entry point failed
[    54.286] (EE) AIGLX: reverting to software rendering
[    54.372] (II) IGLX: enabled GLX_MESA_copy_sub_buffer
[    54.372] (II) IGLX: Loaded and initialized swrast
[    54.372] (II) GLX: Initialized DRISWRAST GL provider for screen 0
[    54.374] (II) intel(0): switch to mode 1920x1080@60.0 on eDP1 using pipe 0, position (0, 0), rotation normal, reflection none
[    54.391] (II) intel(0): Setting screen physical size to 508 x 285
```

**Comment by iprid23** [comment]
Actually there's the qubes-dom0-updates component which includes xorg-x11-drivers. Those drivers are called Xorg DDX, ie. Device Dependent Drivers. To get acceleration you have 2 choices:
Either use the generic modesetting DDX -xorg-x11-modesetting (I know it's stupid that it's called generic modesetting driver and still a DDX, but ...) with glamor or use the i915 DDX from intel(often called just intel) - xorg-x11-drv-intel.

How to do it:
Create /etc/X11/xorg.conf.d/20-intel.conf in dom0 with the following contents:
1. Option 1 - modesetting:
```
Section "device"
    Identifier "intel"
    Driver "modesetting"
    Option "AccelMethod" "glamor"
EndSection
```
2. Option 2 - i915
```
Section "device"
    Identifier "intel"
    Driver "intel"
EndSection
```

Now in option 2, you may choose what acceleration to use:
For example you can add Option "DRI" "2" to disable DRI "3" and have only DRI "2".
Or you can add: Option      "AccelMethod"  "uxa" to use uxa instead of SNA, in the log you've given I see you are using i915 with SNA.

The main benefit of i915 is that you can also add:
    Option "TearFree" "true"

I advise you to have a look at:
https://wiki.archlinux.org/index.php/Intel_Graphics

Also always check the Xorg log.

**Comment by weissjeffm** [comment]
Ok, I tried both of these options, and I still get 
```
(EE) AIGLX: reverting to software rendering
```
with both of them. 

With the former, I get `Screens found but none have a usable configuration`.

With the latter, the Xorg log is the same as my original description of the problem, so I think it was using that driver already.

I didn't quite understand your comment about 
>  there's the qubes-dom0-updates component which includes xorg-x11-drivers.

Was I supposed to install something? I have the package in dom0 `xorg-x11-drv-intel-2.99.917-32.20171025.fc25.x86_64` (which appears to predate my laptop's hardware, with the 2017 date in there).

**Comment by iprid23** [comment]
Section "device"
    Identifier "intel"
    Driver "intel"
    Option "AccelMethod" "uxa"
    Option "DRI" "2"
EndSection

How about this then?

**Comment by weissjeffm** [comment]
@iprid23 I tried that, with those two options, the X server will not start (`Screen(s) found but none have usable configuration`), I had to comment them out (and then I still end up with software render but at least X works).

> As for why I mentioned th qubes-dom0-updates, it's because you can build a newer version yourself if you want with the qubes-builder.

I notice the latest fedora 30 package is `xorg-x11-drv-intel-2.99.917-41.20180618.fc30.x86_64.rpm` (the 2.99.917 part being the same as what I already have) not sure if upgrading to this would make that much difference or maybe the difference between the `-32` build and `-41` matters?

**Comment by iprid23** [comment]
Hm, I can suggest building the module yourself, though it's a bit annoying:
There are two build options you have:
1. Get Fedora 25 VM and use it
2. Use the qubes-builder with chroot

Now there are also two ways to obtain source code:
1. Get a source RPM - ex. http://mirror.in2p3.fr/pub/fedora/linux/releases/30/Everything/source/tree/Packages/x/xorg-x11-drv-intel-2.99.917-41.20180618.fc30.src.rpm
2. Get the source code from this repo:
https://gitlab.freedesktop.org/xorg/driver/xf86-video-intel.git

I think the following is the simplest:
Get a Fedora 25 template, make a standalone build VM based on Fedora 25
Pull this repo https://gitlab.freedesktop.org/xorg/driver/xf86-video-intel.git
Install meson and ninja. Do meson . build. Meson will tell you what dependencies to install.
Do it until it says everything is fine. Then cd build and do ninja. Ninja will build the so library. Do a find -name '*.so', get it and move it dom0. Laziest solution perhaps.

Anyhow, it'd be best if perhaps @marmarek could create a release4.0-fc30 branch in the qubes-linux-dom0-updates, so we can just build the driver from there and also have in the Qubes package repository.

Regards

**Comment by iprid23** [comment]
@weissjeffm I did build the xorg-x11-drv-intel driver from fc30 and that fixed the issue. I can give you RPM packages but then again you should not trust binary packages from me, so I'd advise you to build your own packages or ask some Qubes dev to build them for you.
I will post the patch here so that whoever wants to build it, can do it fast.(though you should read the notes first).

@marmarek - I forked qubes-linux-dom0-updates and was able to build the xorg x11 fc30 packages for fc25, but only nouveau and intel ones. The ati driver had a conflict with libdrm versions.

Notes:
1. Since I was only interested in GPU drivers I did not build the device-mapper-persistent-data here.
So you also need to check if it's okay to use the one from fc30 if you want to use fc30 packages instead.
2. As I said this fails for the ATI rpm driver, so it only works for nouveau and intel.
3. The new xorg-x11-drv-intel driver requires two macros in the fc25 chroot in order to be built:
/home/user/qubes-builder/chroot-dom0-fc25/home/user/.rpmmacros:
%ldconfig_post    %post -p /sbin/ldconfig
%ldconfig_postun  %postun -p /sbin/ldconfig
4. In the patch below I am posting Fedora 30 GPG public key, it's up to whoever uses this to verify that this is the correct key.

```
diff --git a/Makefile b/Makefile
index 013ff29..d673805 100644
--- a/Makefile
+++ b/Makefile
@@ -1,15 +1,15 @@
 FEDORA_MIRROR ?= https://download.fedoraproject.org/pub/fedora/linux
-FC28_BASEURL = $(FEDORA_MIRROR)/releases/28/Everything/source/tree/Packages
-KEYS = keys/RPM-GPG-KEY-fedora-28-primary
+FC30_BASEURL = $(FEDORA_MIRROR)/releases/30/Everything/source/tree/Packages
+KEYS = keys/RPM-GPG-KEY-fedora-30-primary
 
 include Makefile.builder
 
 DIST ?= fc25
 # makefile doesn't have nice function to extract first letter, so use INITIAL
 # placeholder here and fill it using `sed` just before wget invocation
-ALL_URLS := $(addprefix $(FC28_BASEURL)/INITIAL/, $(PACKAGES_FROM_FC28))
+ALL_URLS := $(addprefix $(FC30_BASEURL)/INITIAL/, $(PACKAGES_FROM_FC30))
 
-ALL_FILES := $(notdir $(ALL_URLS:%.fc28.src.rpm=%.$(DIST).src.rpm))
+ALL_FILES := $(notdir $(ALL_URLS:%.fc30.src.rpm=%.$(DIST).src.rpm))
 
 ifneq ($(DISTFILES_MIRROR),)
 ALL_URLS := $(addprefix $(DISTFILES_MIRROR),$(ALL_FILES))
@@ -22,13 +22,13 @@ rpmdb/.imported: $(KEYS)
 	@rpmkeys --dbpath=$(PWD)/rpmdb --import $(KEYS)
 	@touch rpmdb/.imported
 
-%.fc28.src.rpm: rpmdb/.imported
+%.fc30.src.rpm: rpmdb/.imported
 	@wget --no-use-server-timestamps -q -O $@$(UNTRUSTED_SUFF) \
 		$$(echo "$(filter %/$@,$(ALL_URLS))" | sed -e 's:/INITIAL/\(.\):/\1/\1:')
 	@rpmkeys --dbpath=$(PWD)/rpmdb --checksig $@$(UNTRUSTED_SUFF) | grep 'signatures OK$$\|rsa sha1 (md5) pgp md5 OK$$'
 	@mv $@$(UNTRUSTED_SUFF) $@
 
-%.$(DIST).src.rpm: %.fc28.src.rpm
+%.$(DIST).src.rpm: %.fc30.src.rpm
 	mv $< $@
 
 .PHONY: get-sources verify-sources
diff --git a/Makefile.builder b/Makefile.builder
index 86d68f7..6f154a5 100644
--- a/Makefile.builder
+++ b/Makefile.builder
@@ -1,11 +1,10 @@
-PACKAGES_FROM_FC28 := \
-    device-mapper-persistent-data-0.7.5-3.fc28.src.rpm \
-    xorg-x11-drv-intel-2.99.917-32.20171025.fc28.src.rpm \
-    xorg-x11-drv-nouveau-1.0.15-4.fc28.src.rpm \
-    xorg-x11-drv-ati-18.0.1-1.fc28.src.rpm
+PACKAGES_FROM_FC30 := \
+    xorg-x11-drv-intel-2.99.917-41.20180618.fc30.src.rpm \
+    xorg-x11-drv-nouveau-1.0.15-7.fc30.src.rpm \
+    xorg-x11-drv-ati-19.0.1-1.fc30.src.rpm
 
 ifeq ($(shell expr $(subst fc,,$(DIST)) \<= 27 2>/dev/null),1)
-    RPM_SRC_PACKAGES.dom0 := $(PACKAGES_FROM_FC28:%.fc28.src.rpm=%.$(DIST).src.rpm)
+    RPM_SRC_PACKAGES.dom0 := $(PACKAGES_FROM_FC30:%.fc30.src.rpm=%.$(DIST).src.rpm)
 endif
 
 RPM_SRC_PACKAGES := $(RPM_SRC_PACKAGES.$(PACKAGE_SET))
diff --git a/keys/RPM-GPG-KEY-fedora-30-primary b/keys/RPM-GPG-KEY-fedora-30-primary
new file mode 100644
index 0000000..3af51a2
--- /dev/null
+++ b/keys/RPM-GPG-KEY-fedora-30-primary
@@ -0,0 +1,28 @@
+-----BEGIN PGP PUBLIC KEY BLOCK-----
+
+mQINBFturGcBEACv0xBo91V2n0uEC2vh69ywCiSyvUgN/AQH8EZpCVtM7NyjKgKm
+bbY4G3R0M3ir1xXmvUDvK0493/qOiFrjkplvzXFTGpPTi0ypqGgxc5d0ohRA1M75
+L+0AIlXoOgHQ358/c4uO8X0JAA1NYxCkAW1KSJgFJ3RjukrfqSHWthS1d4o8fhHy
+KJKEnirE5hHqB50dafXrBfgZdaOs3C6ppRIePFe2o4vUEapMTCHFw0woQR8Ah4/R
+n7Z9G9Ln+0Cinmy0nbIDiZJ+pgLAXCOWBfDUzcOjDGKvcpoZharA07c0q1/5ojzO
+4F0Fh4g/BUmtrASwHfcIbjHyCSr1j/3Iz883iy07gJY5Yhiuaqmp0o0f9fgHkG53
+2xCU1owmACqaIBNQMukvXRDtB2GJMuKa/asTZDP6R5re+iXs7+s9ohcRRAKGyAyc
+YKIQKcaA+6M8T7/G+TPHZX6HJWqJJiYB+EC2ERblpvq9TPlLguEWcmvjbVc31nyq
+SDoO3ncFWKFmVsbQPTbP+pKUmlLfJwtb5XqxNR5GEXSwVv4I7IqBmJz1MmRafnBZ
+g0FJUtH668GnldO20XbnSVBr820F5SISMXVwCXDXEvGwwiB8Lt8PvqzXnGIFDAu3
+DlQI5sxSqpPVWSyw08ppKT2Tpmy8adiBotLfaCFl2VTHwOae48X2dMPBvQARAQAB
+tDFGZWRvcmEgKDMwKSA8ZmVkb3JhLTMwLXByaW1hcnlAZmVkb3JhcHJvamVjdC5v
+cmc+iQI4BBMBAgAiBQJbbqxnAhsPBgsJCAcDAgYVCAIJCgsEFgIDAQIeAQIXgAAK
+CRDvPBEfz8ZZudTnD/9170LL3nyTVUCFmBjT9wZ4gYnpwtKVPa/pKnxbbS+Bmmac
+g9TrT9pZbqOHrNJLiZ3Zx1Hp+8uxr3Lo6kbYwImLhkOEDrf4aP17HfQ6VYFbQZI8
+f79OFxWJ7si9+3gfzeh9UYFEqOQfzIjLWFyfnas0OnV/P+RMQ1Zr+vPRqO7AR2va
+N9wg+Xl7157dhXPCGYnGMNSoxCbpRs0JNlzvJMuAea5nTTznRaJZtK/xKsqLn51D
+K07k9MHVFXakOH8QtMCUglbwfTfIpO5YRq5imxlWbqsYWVQy1WGJFyW6hWC0+RcJ
+Ox5zGtOfi4/dN+xJ+ibnbyvy/il7Qm+vyFhCYqIPyS5m2UVJUuao3eApE38k78/o
+8aQOTnFQZ+U1Sw+6woFTxjqRQBXlQm2+7Bt3bqGATg4sXXWPbmwdL87Ic+mxn/ml
+SMfQux/5k6iAu1kQhwkO2YJn9eII6HIPkW+2m5N1JsUyJQe4cbtZE5Yh3TRA0dm7
++zoBRfCXkOW4krchbgww/ptVmzMMP7GINJdROrJnsGl5FVeid9qHzV7aZycWSma7
+CxBYB1J8HCbty5NjtD6XMYRrMLxXugvX6Q4NPPH+2NKjzX4SIDejS6JjgrP3KA3O
+pMuo7ZHMfveBngv8yP+ZD/1sS6l+dfExvdaJdOdgFCnp4p3gPbw5+Lv70HrMjA==
+=BfZ/
+-----END PGP PUBLIC KEY BLOCK-----
```

**Comment by weissjeffm** [comment]
@iprid23 that's great you were able to fix the issue, thanks! I am not sure I can build the package myself since I am not sure what you did to build it.

While technically I shouldn't trust your binary, the odds of it being malicious are so low that I won't worry about it (I mean, what kind of attacker goes after a single user who just happened to open a github issue?). If there's an easy way for me to build it myself, I'll do that, but otherwise I don't mind taking a binary. I have a fedora-25 vm with all the dev dependencies installed as per your previous comment, and I've cloned the `qubes-linux-dom0-updates` repo but I'm not sure how to build the packages. I tried to apply the patch but it wouldn't apply:

```
[user@intel-driver-build qubes-linux-dom0-updates]$ git apply ~/patch
error: patch failed: Makefile:1
error: Makefile: patch does not apply
error: patch failed: Makefile.builder:1
error: Makefile.builder: patch does not apply
[user@intel-driver-build qubes-linux-dom0-updates]$ 
```

**Comment by iprid23** [comment]
@weissjeffm For some reason the fc25 packages for dom0 are using a custom pathed gcc. So instead of trying to use a vm with fc25 and doing make, I'd advise you to actually use the qubes-builder. If however you prefer not to bother with that, you could technically build in dom0, though I see that as a security risk too. Instead I advise you to try the standard way:

So implying you use a vm based on fc30 template:
Pull qubes-builder, then you need to setup the GPG keys and install dev dependecies. It's a bit tricky, because some of the qubes-builder docs are obsolete.
I will give you a quick idea of how I do it:
Pre-reqs:
VM must have at least 50 GB of space and passwordless sudo(it's the default behaviour), 8 GB ram is preferable for good speed, 2 CPUs is fine, more is better
So in the fc30 vm:
1. Build deps
sudo dnf install perl-Digest-SHA rpmdevtools rpm-build dialog rpm-sign python2-sh dpkg-dev debootstrap devscripts
2. Builder repo
git clone git://github.com/QubesOS/qubes-builder.git
3. GPG key setup
export GNUPGHOME=~/qubes-builder/keyrings/git
mkdir --parents "$GNUPGHOME"
chmod --recursive 700 "$GNUPGHOME"
4. GPG key download, import and trust
wget https://keys.qubes-os.org/keys/qubes-master-signing-key.asc
gpg --import qubes-master-signing-key.asc 
gpg --edit-key 36879494
fpr
# Verify fingerprint! See Note below!
# Once verified, set trust to *ultimate*
# (Typical sequence is trust, 5, q)
wget https://keys.qubes-os.org/keys/qubes-developers-keys.asc
gpg --import qubes-developers-keys.asc
5. Repo verify with GPG keys:
cd qubes-builder
git tag -v `git describe`
6. Actual building:
cp example-configs/qubes-os-r4.0.conf builder.conf
Now you have two choices:
Build gcc from source(can take 6 or 12 hours since it uses only one CPU) or take gcc from Qubes repository. Building is simpler though because you do not need to configure anything.
Just do make gcc-dom0. This will build gcc.
Afterwards do make linux-dom0-updates. It will may ask you to build other things beforehand, if so just do ```make thing``` until it works.
7. Now that we have built the old linux-dom0-updates, apply the patch I gave here. And do not forget to reread my notes regarding the need to add the .rpmmacros file.
8. Just do make COMPONENTS="qubes-linux-dom0-updates" get-sources-extra
 This will download the srpms for the new drivers.
After that just do make qubes-linux-dom0-updates
This should build the new drivers and you should find them in: 
/home/user/qubes-builder/qubes-src/linux-dom0-updates/pkgs/dom0-fc25/x86_64

If you mess up on the way, don't worry, many tutorials/infos are obsolete or contain small mistakes. I am trying to help people build Qubes as to make it more popular(took me a lot of time to learn to build it), so feel free to ask me or some of the qubes devs. @marmarek is always helpful

**Comment by weissjeffm** [comment]
@iprid23 thanks so much for the detailed steps!

I got as far as running `make linux-dom0-updates` which fails after about 30 seconds, and this is the end of the log (below). I don't see anything in the log where it's asking me to build other packages first. Any idea what's wrong? 

```
Processing files: device-mapper-persistent-data-0.7.5-3.fc25.x86_64
Executing(%doc): /bin/sh -e /var/tmp/rpm-tmp.2iOWO7
+ umask 022
+ cd /home/user/rpmbuild/BUILD
+ cd thin-provisioning-tools-0.7.5
+ DOCDIR=/home/user/rpmbuild/BUILDROOT/device-mapper-persistent-data-0.7.5-3.fc25.x86_64/usr/share/doc/device-mapper-persistent-data
+ export DOCDIR
+ /usr/bin/mkdir -p /home/user/rpmbuild/BUILDROOT/device-mapper-persistent-data-0.7.5-3.fc25.x86_64/usr/share/doc/device-mapper-persistent-data
+ cp -pr COPYING /home/user/rpmbuild/BUILDROOT/device-mapper-persistent-data-0.7.5-3.fc25.x86_64/usr/share/doc/device-mapper-persistent-data
+ cp -pr README.md /home/user/rpmbuild/BUILDROOT/device-mapper-persistent-data-0.7.5-3.fc25.x86_64/usr/share/doc/device-mapper-persistent-data
+ exit 0
Provides: device-mapper-persistent-data = 0.7.5-3.fc25 device-mapper-persistent-data(x86-64) = 0.7.5-3.fc25
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Requires: libaio.so.1()(64bit) libaio.so.1(LIBAIO_0.1)(64bit) libaio.so.1(LIBAIO_0.4)(64bit) libc.so.6()(64bit) libc.so.6(GLIBC_2.14)(64bit) libc.so.6(GLIBC_2.2.5)(64bit) libc.so.6(GLIBC_2.3.4)(64bit) libc.so.6(GLIBC_2.4)(64bit) libc.so.6(GLIBC_2.7)(64bit) libdl.so.2()(64bit) libdl.so.2(GLIBC_2.2.5)(64bit) libexpat.so.1()(64bit) libgcc_s.so.1()(64bit) libgcc_s.so.1(GCC_3.0)(64bit) libm.so.6()(64bit) libm.so.6(GLIBC_2.2.5)(64bit) libstdc++.so.6()(64bit) libstdc++.so.6(CXXABI_1.3)(64bit) libstdc++.so.6(GLIBCXX_3.4)(64bit) libstdc++.so.6(GLIBCXX_3.4.11)(64bit) libstdc++.so.6(GLIBCXX_3.4.15)(64bit) libstdc++.so.6(GLIBCXX_3.4.20)(64bit) libstdc++.so.6(GLIBCXX_3.4.21)(64bit) libstdc++.so.6(GLIBCXX_3.4.9)(64bit) rtld(GNU_HASH)
Processing files: device-mapper-persistent-data-debuginfo-0.7.5-3.fc25.x86_64
Provides: device-mapper-persistent-data-debuginfo = 0.7.5-3.fc25 device-mapper-persistent-data-debuginfo(x86-64) = 0.7.5-3.fc25
Requires(rpmlib): rpmlib(CompressedFileNames) <= 3.0.4-1 rpmlib(FileDigests) <= 4.6.0-1 rpmlib(PayloadFilesHavePrefix) <= 4.0-1
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/user/rpmbuild/BUILDROOT/device-mapper-persistent-data-0.7.5-3.fc25.x86_64
Wrote: pkgs/dom0-fc25/x86_64/device-mapper-persistent-data-0.7.5-3.fc25.x86_64.rpm
Wrote: pkgs/dom0-fc25/x86_64/device-mapper-persistent-data-debuginfo-0.7.5-3.fc25.x86_64.rpm
Executing(%clean): /bin/sh -e /var/tmp/rpm-tmp.CmbFTt
+ umask 022
+ cd /home/user/rpmbuild/BUILD
+ cd thin-provisioning-tools-0.7.5
+ exit 0
Executing(--clean): /bin/sh -e /var/tmp/rpm-tmp.pSs0YP
+ umask 022
+ cd /home/user/rpmbuild/BUILD
+ rm -rf thin-provisioning-tools-0.7.5
+ exit 0
make[2]: Leaving directory '/home/jweiss/qubes-builder'
make[2]: Entering directory '/home/jweiss/qubes-builder'
make[2]: *** No rule to make target '/home/jweiss/qubes-builder/chroot-dom0-fc25/home/user/qubes-src/linux-dom0-updates/xorg-x11-drv-intel-2.99.917-32.20171025.fc25.src.rpm', needed by 'dist-build-dep'.  Stop.
make[2]: Leaving directory '/home/jweiss/qubes-builder'
```

**Comment by weissjeffm** [comment]
I think the issue is there's no fedora28 gpg keys in the chroot. I'm trying to figure out how to get them in there.

**Comment by weissjeffm** [comment]
OK i was able to build the packages now. I think what was missing was: `make COMPONENTS="qubes-linux-dom0-updates" get-sources-extra` before running `make linux-dom0-updates` the first time. I ended up blundering around until I ran "make get-sources" in the `linux-dom0-updates` dir and then i could run `make linux-dom0-updates` at the top level.

So then I tarred up the rpms, copied them to dom0 and installed them with `dnf`. Rebooted. Unfortunately it did not fix the issue. Exact same problems as before, although now I get this warning that I didn't used to:

```
(WW) intel(0): Unknown chipset
```
the package i installed is `xorg-x11-drv-intel-2.99.917-41.20180618.fc25.x86_64`. I am not sure if others needed to be installed, but I also added `device-mapper-persistent-data-0.7.5-3.fc25.x86_64`. (no new dependencies were required for either of them)

**Comment by rysiekpl** [comment]
I was experiencing the same issue on a Thinkpad T470 (i5-7200U). This helped:

```
Section "Device"
    Identifier "intel"
    Driver "modesetting"
    Option "AccelMethod" "glamor"
    Option "DRI" "3"
EndSection
```

Now I got a 3D-accelerated desktop without artifacts. More context: 

 - https://github.com/QubesOS/qubes-issues/issues/4782
 - https://groups.google.com/d/msgid/qubes-users/5cc49553-b12c-4e4d-7601-f961330a14e6%40gmail.com

**Comment by marmarek** [comment]
@iprid23 can you verify if after the update it's still intel driver in use, not modesetting one?

**Comment by scsich** [comment]
> @iprid23 can you verify if after the update it's still intel driver in use, not modesetting one?

I followed this path but with the same PCI id as @weissjeffm. Thanks to @iprid23 I was able to build xorg-intel driver from the FC30 source repos he suggested. However, I discovered that the Whiskey lake PCI id of 0x3ea0 isn't in that source repo so I had to unpack/repack source RPM and add the attached patch.

Also, had to patch libdrm to add 0x3ea0 and build that src rpm.

I tried both Intel and mode setting routes with intel yielding:

```
[ 13691.400] (EE) AIGLX error: Calling driver entry point failed
[ 13691.401] (EE) AIGLX: reverting to software rendering
[ 13696.275] (EE) intel(0): Failed to submit rendering commands (Bad address), disabling acceleration.
```

and modesetting yielding

```
[ 14228.463] (II) glamor: EGL version 1.4 (DRI2):
[ 14228.463] EGL_MESA_drm_image required.
[ 14228.464] (EE) modeset(0): glamor initialization failed
[ 14228.614] (EE) AIGLX: reverting to software rendering

```

I then went down the rabbit hole of trying to get Mesa 18.3+ built (19 out of fc30 repos) but llvm, meson, and clang inside the fc25 qubes builder chroot are too old. Meson was easy to solve but llvm...

https://www.phoronix.com/scan.php?page=news_item&px=Mesa-18.3-Whiskey-Lake-Added

@marmarek is there some trick to getting an updated mesa in fc25 dom0 that doesn't involve a from-source install of llvm/clang 7+ in chroot?

**Comment by marmarek** [comment]
> @marmarek is there some trick to getting an updated mesa in fc25 dom0 that doesn't involve a from-source install of llvm/clang 7+ in chroot?

I'm afraid not really :(

**Comment by rysiekpl** [comment]
I am experiencing the same issue on a T490 (and have been since I got it, for about a year and a half). My previous `modesetting` fix that I used for my T470 does not work, because the Intel graphics chip used in T490, while [effectively the same as in T470](https://en.wikichip.org/wiki/intel/uhd_graphics/620), uses a different PCIID (`3EA0` in my case).

Literally the only fix that is needed is adding this PCIID to `xf86-video-intel-20180618/src/i915_pciids.h`.

Is there any chance we get an updated Mesa in Qubes 4.0? Clearly there is quite a few people having that problem. Intel (U)HD line is used in a lot of supported laptops, and there are not many of those in general: twice I was choosing a laptop, and twice I chose hardware *specifically* so that it can run Qubes; and one of these times the exact same graphics chipset just doesn't get full support.

I appreciate that Qubes does not have the resources to support all hardware, and that work on the R4.1 takes precedence over fixing some old issues in R4.0. However, there is something to be said for bringing full support to not-that-bleeding-edge hardware that is marked as well-supported, especially if it just boils down to a few lines patch and rebuild of Mesa.

(and yes, I will probably attempt such a patch-and-rebuild myself, but even if I succeed it's just me; it would be wonderful if others could benefit from it...)
